### DISCUSSION TITLE
Wrong modifiers parsing when using value in custom directive

### DISCUSSION THREAD

**Discussion by leoboyerbx** [alpinejs/alpine#3585] [7. Bugs (archived)]
Hello, I think I found a bug with the extending system of Alpine.

When defining a custom directive, if the element has both modifiers and a value, the last modifier gets parsed with the value and is not extracted.

I have done a reproduction of this bug on [Stackblitz](https://stackblitz.com/edit/web-platform-rzrl7r?file=index.html):

```html
<div x-data x-foo.modifier1.modifier2:value="expression">
      Open console to see bug
    </div>

    <script>
      document.addEventListener('alpine:init', () => {
        Alpine.directive('foo', (el, { value, expression, modifiers }) => {
          console.log(expression, value, modifiers);
        });
      });
    </script>
```

Expected log:
```javascript
expression value ["modifier1", "modifier2"]
```
Actual log:
```javascript
expression value ["modifier1", "modifier2:value"]
```

**Comment by leoboyerbx** [comment]
We should update the regex parsing the modifiers from `\.[^.\]]+(?=[^\]]*$)` to `\.[^:.\]]+(?=[^\]]*$)`

**Comment by ekwoka** [comment]
change

`x-foo.modifier1.modifier2:value`

to

`x-foo:value.modifier1.modifier2`

**Reply by trych** [reply]
tbh it's not very obvious from the docs that the order needs to be strictly `:value.modifiers`.

**Reply by ekwoka** [reply]
Sure, could definitely be improved in the documentation. I think it's missing a few things.

But it is a pattern.

Conceptually it doesn't make any sense for the modifiers to be after the value anyway.

And the modifiers modify the thing as defined by the value.
